### PR TITLE
[DIC][3.3] ServiceLocator generated with invalid callable

### PR DESCRIPTION
**Pull request by linaori**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | ~
| License       | MIT
| Doc PR        | ~

Currently, when having a service definition which has more than 1 usage of that service, it will put the argument in `$a` for example. By doing so, it will also use `$a` in the callable of the `ServiceLocator`, which result in an undefined variable `$a`.

I managed to trigger this with the translator service, where I have my own loader, that I add for NL and EN, causing 2 method calls to turn the direct getter into a variable. I've added 2 test cases, 1 with only 1 method call and 1 with 2 method calls.

I'm not sure if this is the correct solution, but this managed to make my application work again with 3.3.

ping @nicolas-grekas

### CODE REVIEW

**Comment by stof** [comment]
This is not good. Service locators are about making instantiation lazy, so they should not rely on services instantiated early when instantiating the ServiceLocator itself. We should prevent inlining and reusing the variables in this case actually

File: src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php [hunk 1]
```diff
@@ -533,6 +534,20 @@ public function testServiceLocator()
...
+        // 1 callable method
+        $container->register('service_locator_with_inline_reference_1', ServiceLocator::class)
+            ->addArgument(array('method_call_class' => new ServiceClosureArgument(new Reference('method_call_class'))))
+            ->addMethodCall('callableMethod', array('a', new Reference('method_call_class')));
+
+        // 2 callable methods
+        $container->register('service_locator_with_inline_reference_2', ServiceLocator::class)
+            ->addArgument(array('method_call_class' => new ServiceClosureArgument(new Reference('method_call_class'))))
+            ->addMethodCall('callableMethod', array('a', new Reference('method_call_class')))
+            ->addMethodCall('callableMethod', array('b', new Reference('method_call_class')));
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php]: Calling methods on a service locator looks weird to me. It breaks the laziness of the locator (as instantiating the ServiceLocator involves getting some instances in it to call the method). The method call should be added in the definition of the referenced service instead.
I don't see any use case for such calling on the ServiceLocator instance (and so not sure we should allow such feature and spend time fixing edge cases in it).

/cc @nicolas-grekas

**Comment by iltar** [comment]
> so they should not rely on services instantiated early when instantiating the ServiceLocator itself

They have to be instantiated anyway, due to the method calls, so that doesn't matter. I agree that this solution is not the best, but I couldn't quite figure out where this was calculated.

Would you rather see this for the 2 method calls?

```php
    /**
     * Gets the 'service_locator_with_inline_reference_2' service.
     *
     * This service is shared.
     * This method always returns the same instance of the service.
     *
     * @return \Symfony\Component\DependencyInjection\ServiceLocator A Symfony\Component\DependencyInjection\ServiceLocator instance
     */
    protected function getServiceLocatorWithInlineReference2Service()
    {
        $this->services['service_locator_with_inline_reference_2'] = $instance = new \Symfony\Component\DependencyInjection\ServiceLocator(array('method_call_class' => function () use ($a) {
            return ${($_ = isset($this->services['method_call_class']) ? $this->services['method_call_class'] : $this->get('method_call_class')) && false ?: '_'};
        }));

        $instance->callableMethod('a', ${($_ = isset($this->services['method_call_class']) ? $this->services['method_call_class'] : $this->get('method_call_class')) && false ?: '_'});
        $instance->callableMethod('b', ${($_ = isset($this->services['method_call_class']) ? $this->services['method_call_class'] : $this->get('method_call_class')) && false ?: '_'});

        return $instance;
    }
```

Not sure if this will have a performance impact.

**Comment by stof** [comment]
@iltar see my other comment. I would rather remove the ability to perform magic calls on the ServiceLocator entirely (unless we have a use case for it that I missed)

File: src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php (hunk 1 above)
Comment (iltar) [inline src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php]: This is the compiler pass that triggers this edge case:

```php
final class TranslationLoaderPass implements CompilerPassInterface
{
    /**
     * {@inheritdoc}
     */
    public function process(ContainerBuilder $container)
    {
        $loader     = new Reference('cms_api.translations.loader');
        $translator = $container->findDefinition('translator');
        $locales    = $this->findLocales($container, $translator);

        foreach ($locales as $locale) {
            $translator->addMethodCall('addResource', ['cms_api', $loader, $locale]);
        }
    }

    /**
     * @param ContainerBuilder $container
     * @param Definition       $translator
     * @return string[]
     */
    private function findLocales(ContainerBuilder $container, Definition $translator)
    {
        $locales = [$container->getParameter('kernel.default_locale')];
        foreach ($translator->getMethodCalls() as list($method, $args)) {
            if ('setFallbackLocales' !== $method) {
                continue;
            }

            $locales = array_merge($locales, $args[0]);
        }

        return array_unique($locales);
    }
}

```

**Comment by iltar** [comment]
Here's a snippet of the generated container:
```php
$a = ${($_ = isset($this->services['cms_api.translations.loader']) ? $this->services['cms_api.translations.loader'] : $this->get('cms_api.translations.loader')) && false ?: '_'};

$this->services['translator.default'] = $instance = new \Symfony\Bundle\FrameworkBundle\Translation\Translator(new \Symfony\Component\DependencyInjection\ServiceLocator(array('cms_api.translations.loader' => function () {
    return $a;
}, 'hostnet_entity_translation.enum.loader' => function () {
    return ${($_ = isset($this->services['hostnet_entity_translation.enum.loader']) ? $this->services['hostnet_entity_translation.enum.loader'] : $this->get('hostnet_entity_translation.enum.loader')) && false ?: '_'};
}, 'translation.loader.csv' => function () {
    return ${($_ = isset($this->services['translation.loader.csv']) ? $this->services['translation.loader.csv'] : $this->get('translation.loader.csv')) && false ?: '_'};
},
    // ...

$instance->addResource('enum', /* ... */, 'en', /* ... */);
$instance->addResource('enum', /* ... */, 'en', /* ... */);
$instance->addResource('cms_api', $a, 'nl');
$instance->addResource('cms_api', $a, 'en');
```

```php
final class TranslationLoaderPass implements CompilerPassInterface
{
    /**
     * {@inheritdoc}
     */
    public function process(ContainerBuilder $container)
    {
        $loader     = new Reference('cms_api.translations.loader');
        $translator = $container->findDefinition('translator');
        $locales    = $this->findLocales($container, $translator);

        foreach ($locales as $locale) {
            $translator->addMethodCall('addResource', ['cms_api', $loader, $locale]);
        }
    }

    /**
     * @param ContainerBuilder $container
     * @param Definition       $translator
     * @return string[]
     */
    private function findLocales(ContainerBuilder $container, Definition $translator)
    {
        $locales = [$container->getParameter('kernel.default_locale')];
        foreach ($translator->getMethodCalls() as list($method, $args)) {
            if ('setFallbackLocales' !== $method) {
                continue;
            }

            $locales = array_merge($locales, $args[0]);
        }

        return array_unique($locales);
    }
}
```

Here's the service XML (minus a few resources):

```xml
    <service id="translator.default" class="Symfony\Bundle\FrameworkBundle\Translation\Translator">
      <argument type="service">
        <service class="Symfony\Component\DependencyInjection\ServiceLocator" public="false">
          <tag name="container.service_locator"/>
          <argument type="collection">
            <argument key="cms_api.translations.loader" type="service" id="cms_api.translations.loader"/>
            <argument key="hostnet_entity_translation.enum.loader" type="service" id="hostnet_entity_translation.enum.loader"/>
            <argument key="translation.loader.csv" type="service" id="translation.loader.csv"/>
            <argument key="translation.loader.dat" type="service" id="translation.loader.dat"/>
            <argument key="translation.loader.ini" type="service" id="translation.loader.ini"/>
            <argument key="translation.loader.json" type="service" id="translation.loader.json"/>
            <argument key="translation.loader.mo" type="service" id="translation.loader.mo"/>
            <argument key="translation.loader.php" type="service" id="translation.loader.php"/>
            <argument key="translation.loader.po" type="service" id="translation.loader.po"/>
            <argument key="translation.loader.qt" type="service" id="translation.loader.qt"/>
            <argument key="translation.loader.res" type="service" id="translation.loader.res"/>
            <argument key="translation.loader.xliff" type="service" id="translation.loader.xliff"/>
            <argument key="translation.loader.yml" type="service" id="translation.loader.yml"/>
          </argument>
        </service>
      </argument>
      <argument type="service">
        <service class="Symfony\Component\Translation\MessageSelector" public="false"/>
      </argument>
      <argument>nl</argument>
      <argument type="collection">
        <argument key="hostnet_entity_translation.enum.loader" type="collection">
          <argument>enum</argument>
        </argument>
        <argument key="translation.loader.php" type="collection">
          <argument>php</argument>
        </argument>
        <argument key="translation.loader.yml" type="collection">
          <argument>yml</argument>
        </argument>
        <argument key="translation.loader.xliff" type="collection">
          <argument>xlf</argument>
          <argument>xliff</argument>
        </argument>
        <argument key="translation.loader.po" type="collection">
          <argument>po</argument>
        </argument>
        <argument key="translation.loader.mo" type="collection">
          <argument>mo</argument>
        </argument>
        <argument key="translation.loader.qt" type="collection">
          <argument>ts</argument>
        </argument>
        <argument key="translation.loader.csv" type="collection">
          <argument>csv</argument>
        </argument>
        <argument key="translation.loader.res" type="collection">
          <argument>res</argument>
        </argument>
        <argument key="translation.loader.dat" type="collection">
          <argument>dat</argument>
        </argument>
        <argument key="translation.loader.ini" type="collection">
          <argument>ini</argument>
        </argument>
        <argument key="translation.loader.json" type="collection">
          <argument>json</argument>
        </argument>
        <argument key="cms_api.translations.loader" type="collection">
          <argument>cms_api</argument>
        </argument>
      </argument>
      <argument type="collection">
        <argument key="cache_dir">/home/ivanderberg/projects/mijn/www/var/cache/dev/translations</argument>
        <argument key="debug">true</argument>
        <argument key="resource_files" type="collection" />
      </argument>
      <call method="setConfigCacheFactory">
        <argument type="service" id="config_cache_factory"/>
      </call>
      <call method="setFallbackLocales">
        <argument type="collection">
          <argument>en</argument>
        </argument>
      </call>
      <call method="addResource">
        <argument>enum</argument>
        <argument>...</argument>
        <argument>en</argument>
        <argument>...</argument>
      </call>
      <call method="addResource">
        <argument>enum</argument>
        <argument>...</argument>
        <argument>en</argument>
        <argument>...</argument>
      </call>
      <call method="addResource">
        <argument>cms_api</argument>
        <argument type="service" id="cms_api.translations.loader"/>
        <argument>nl</argument>
      </call>
      <call method="addResource">
        <argument>cms_api</argument>
        <argument type="service" id="cms_api.translations.loader"/>
        <argument>en</argument>
      </call>
    </service>
```

**Comment by stof** [comment]
Well, OK. This is involving method calls on the class in which the ServiceLocator is injected, not on the ServiceLocator itself. This is a valid use case.
So it means we still need to fix the issue.

However, I don't like the implementation doing a preg_match on the code. It is too fragile IMO.

@nicolas-grekas 2 questions here:

- should we use a solution based on `use ($a)` in the closure here, or force the service locator to read it from the container again ? Is there any benefit using `use ($a)` (which IMO is more complex to implement in a clean way) ?
- what is the use case for the magic calls on the ServiceLocator ? Could we remove it from the 3.3 API ? I don't see a use case for it, as it forces breaking lazyness, and so should never be used in a service configuration in the container IMO.

**Comment by stof** [comment]
@iltar please rewrite your test to rely on a real-word use case: the method call being on a service in which the ServiceLocator is inlined, rather than being on the ServiceLocator itself.

**Comment by nicolas-grekas** [comment]
We should only allow references at this point, thus I think PhpDumper does need no patch. Instead, we should do this (and fix any reasons for this exception to be triggered in a non-buggy config).
Does that make sense?

```diff
--- a/src/Symfony/Component/DependencyInjection/Argument/ClosureProxyArgument.php
+++ b/src/Symfony/Component/DependencyInjection/Argument/ClosureProxyArgument.php
@@ -12,6 +12,7 @@
 namespace Symfony\Component\DependencyInjection\Argument;
 
 use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\DependencyInjection\Exception\InvalidArgumentException;
 use Symfony\Component\DependencyInjection\Reference;
 
 /**
@@ -41,6 +42,9 @@ class ClosureProxyArgument implements ArgumentInterface
      */
     public function setValues(array $values)
     {
+        if (!$values[0] instanceof Reference) {
+            throw new InvalidArgumentException(sprintf('A ClosureProxyArgument must hold a Reference, "%s" given.', is_object($values[0]) ? get_class($values[0]) : gettype($values[0])));
+        }
         list($this->reference, $this->method) = $values;
     }
 }
```

**Comment by stof** [comment]
@nicolas-grekas we need patching: dumping the Reference inside the locator should either not rely on variables defined in the main function, or they need to be imported in the closure

**Comment by iltar** [comment]
I'll update the tests tomorrow to see what I can do, shouldn't be hard to create a real-world example based off the translator case I have. What do you recommend instead of the `preg_match`? I know it's not the best way to do so, but in this case, variables should always be made in the same format right? I don't think it's easy to get the code given the `$a`, to give it a different way, because the inlined variant is not detected as second usage either (see test case with 1 method call). It only starts failing when more than 1 usage is present.

**Comment by nicolas-grekas** [comment]
Hum, OK. Instead of trying to add a `use`, what about finding a way to NOT enter this `if`:
https://github.com/symfony/symfony/blob/master/src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php#L1527

this is the culprit.
@iltar please also take the patch above. It's not directly related, but it's still useful as a safeguard.

**Comment by stof** [comment]
@nicolas-grekas we should temporarily reset the list of declaration variables while dumping a ServiceLocator or a ClosureProxyArgument

**Comment by nicolas-grekas** [comment]
@stof yep, I think we should do that for any `ArgumentInterface` in fact, that's aligned to its semantics.

**Comment by stof** [comment]
@nicolas-grekas true

**Comment by iltar** [comment]
I've updated the examples to represent the real world scenario I am facing. If I understand it correct, the desired behavior would reflect this patch?

```diff
diff --git a/src/Symfony/Component/DependencyInjection/Tests/Fixtures/php/services_locator.php b/src/Symfony/Component/DependencyInjection/Tests/Fixtures/php/services_locator.php
index 7b263fa..d623194 100644
--- a/src/Symfony/Component/DependencyInjection/Tests/Fixtures/php/services_locator.php
+++ b/src/Symfony/Component/DependencyInjection/Tests/Fixtures/php/services_locator.php
@@ -188,8 +188,8 @@ class ProjectServiceContainer extends Container
     {
         $a = ${($_ = isset($this->services['translator.loader_3']) ? $this->services['translator.loader_3'] : $this->get('translator.loader_3')) && false ?: '_'};
 
-        $this->services['translator_3'] = $instance = new \Symfony\Component\DependencyInjection\Tests\Fixtures\StubbedTranslator(new \Symfony\Component\DependencyInjection\ServiceLocator(array('translator.loader_3' => function () use ($a) {
-            return $a;
+        $this->services['translator_3'] = $instance = new \Symfony\Component\DependencyInjection\Tests\Fixtures\StubbedTranslator(new \Symfony\Component\DependencyInjection\ServiceLocator(array('translator.loader_3' => function () {
+            return ${($_ = isset($this->services['translator.loader_3']) ? $this->services['translator.loader_3'] : $this->get('translator.loader_3')) && false ?: '_'};
         })));
 
         $instance->addResource('db', $a, 'nl');
```

While I understand the need of lazyness, the service is _already_ instantiated due to the method calls, regardless of having a `use ($a)` or that long `${}` notation.

**Comment by nicolas-grekas** [comment]
@iltar that's true, but that's an unneeded micro optimization to me: the internal use of the long notation is marginally different. I guess the issue also exists with the other `ArgumentInterface` subtypes.
Logically, it's important also to me to split scopes. having this `use` breaks the scope separation.
So, in short: let's KISS :)

**Comment by iltar** [comment]
Fair enough. Do you have any pointers on how to detect whether I'm in a different scope?

**Comment by nicolas-grekas** [comment]
there is none, the check is just `$value instanceof ArgumenInterface`

**Comment by iltar** [comment]
@nicolas-grekas is this what you had in mind? I've removed the whole temporary variable feature, which seems a lot easier. 

_On a side note, any particular reason `dumpValue` has `elseif`/`else` spread everywhere instead of guard clauses? I can imagine for merging reasons in lower branches, but it would read a lot easier_

**Comment by nicolas-grekas** [comment]
getServiceCallsFromArguments is not used anymore, could be removed

> instead of guard clauses

we should not change that. There is one reason: history. But we discourage refactoring existing for pure CS purposes.

**Comment by stof** [comment]
I fear that removing the reference variables stuff entirely could create issues related to cases where services have circular references which are partially inlined, as this change the order of instantiation.

And this must be benchmarked as well, because I think that storing references used multiple times in a local variables was introduced for performance reason. So we must ensure that we don't loose performance at runtime here.

**Comment by iltar** [comment]
@stof maybe we could add some test-cases for that?

**Comment by nicolas-grekas** [comment]
Thank you @iltar, I finished this fix in #22581.
